### PR TITLE
Add a minimal base image

### PR DESCRIPTION
**Pull request by staroselskii**
Hello, guys!

In this PR I propose to use a edison-image-minimal as a base for other images. This seems to be the cleanest solution for application meta-layers to depend on.

These layers then would have for example another my-awesome-application.bb image recipe with a `require recipes-core/images/edison-image-minimal.bb`. This has worked quite good for me for a while.

I'm eager to hear your thoughts.

### CODE REVIEW

**Comment by htot** [comment]
Good idea, question is of course: what goes into minimal? Just boots? Sets up bt,wifi/eth? I'll have a look at your partitioning maybe this weekend. And would be good to hear @alext-mkrs opinion.

**Comment by staroselskii** [comment]
@htot 

I should've actually name it an RFC, because I'm not sure what should go into minimal. I just left the stuff that I use on almost every boot. I guess swig and tzdata should also go from minimal to the full image.

cleanjournal is kind of a mystery for me. It seems like it is responsible for some weird "bisect journal I/O errors" that I sometimes see.

So, once again: your mileage may vary so we should discuss and I'm happy to update the PR accordingly.

**Comment by staroselskii** [comment]
@htot 

I think that btrfs is certainly something that we shouldn't mix with this PR.

**Comment by htot** [comment]
So what is the minimum that the edison-image-minimal should support?

I my mind: not the tools that are used for debugging (htop, iotop, systemd-analyze etc). Maybe even not the stuff that are nice to have for a user that has cli access (timezone, less, screen). 

But I think at least stuff that gets hardware initialized (oobe - when I get that to work, related python and nodejs and daemons, ap-mode-toggle etc). Probably should trigger initramfs/kernel + U-Boot build too. This way you get autoconfigured or web server configured device that doesn't do much else.

Then edison-image should add everything a developer (with Edison-Arduino) wants (cli tools not used by scripts/deamons, debugging / perf related). 

While specific applications would have there own layer with image including edison-image-minimal and adding whatever else needed?

**Comment by alext-mkrs** [comment]
The idea LGTM. For the minimal one I'd suggest anything and everything that gets the thing booted and all the comms interfaces up (BT, WiFi, USB). Essentially what @htot proposed as well.

I'm not sure I get the idea of applications having the whole image recipe included - are we talking about single-app images for Edison where that single app's recipe would tweak the whole OS image to suit the needs and cut everything else? Essentially, the proposed split is okay even without that - I like modularity it introduces :)

Just one nitpick - the commit message in the first one ("distro: core: images: add edison-image-minimal") has several typos, I'd suggest to correct them before merging.

**Comment by staroselskii** [comment]
@alext-mkrs Wow, these are some serious typos! Sorry for that! And thanks for the review. I'll force push the changes, then?

@htot 

> But I think at least stuff that gets hardware initialized (oobe - when I get that to work, related python and nodejs and daemons, ap-mode-toggle etc). Probably should trigger initramfs/kernel + U-Boot build too. This way you get autoconfigured or web server configured device that doesn't do much else.

Are you sure nodejs is needed?

So, just to be on one page: do I get rid of less,screen,strace et al.?

I'm also not sure about swig? It seems more like a build dependency rather than something we'd like to have on the device, doesn't it? (I've had only a very limited SWIG exposure so there's a chance I don't understand all the detail).

**Comment by htot** [comment]
I think we are on the same page.

Currently I'm fixing up the `oobe` which consists of a python configuration script and a `nodejs` web server. This thread made me think I can integrate the web server in the python script. And afaik other than that  `nodejs` is only used with `mraa` and `upm` which might be moved out the base image then.

I don't know why we have `screen`, but ``strace should go I think. And `less` I added because in interactive mode `journalctl` looks messed because of the `busybox` replacement.

**Comment by staroselskii** [comment]
@htot 

It'd be nice if nodejs could be left out of the image.

> And less I added because in interactive mode `journalctl` looks messed because of the `busybox` replacement.
I propose to leave it be, then! Why suffer?

> I don't know why we have screen
I usually use screen when deal with UART, so this is kind of hardware related.

> Currently I'm fixing up the oobe

As a matter of fact I haven't used it, so can't comment on that.

**Comment by htot** [comment]
> > Currently I'm fixing up the oobe
> As a matter of fact I haven't used it, so can't comment on that.
It is a script used to setup the machine name, root password, wifi connection and used from the command line. It then doesn't do much you couldn't do in another way, but like a configuration wizard makes life easier. When called by the web server, it switches to AP mode, gets a list of available AP's, lets you enter the password and then connects to that AP. So you can potentially configure from a laptop or phone over wifi.

**Comment by staroselskii** [comment]
@htot 

IMHO, this is definitely something that shouldn't go in the minimal image because, well, we can do it using some underlying tools.

**Comment by staroselskii** [comment]
I force-pushed the changes.

Moved to edison-image.bb:

- strace
- iotop and powertop
- swig

**Comment by staroselskii** [comment]
@htot 

Thanks for the review! A bit of sloppiness on my side. I'll do a force push tomorrow.

File: meta-intel-edison-distro/recipes-core/images/edison-image.bb [hunk 1]
```diff
@@ -1,36 +1,9 @@
...
-IMAGE_FSTYPES = "ext4 live"
-
-inherit core-image
-
-# image-live/live-vm-common do not allow to use initramfs kernel
-# and we really want to have one in the generated hddimg file.
-# Otherwise it doesn't boot due to no mmc/sdhc modules being built-in in our setup).
-# There seems to be no standard mechanism for doing that, but the below works.
+DESCRIPTION = "A fully functional image to run EDISON"
 KERNEL_IMAGETYPE_pn-edison-image = "bzImage-initramfs-edison.bin"
```
[the remark this answers is not in the document]
Comment (staroselskii) [reply meta-intel-edison-distro/recipes-core/images/edison-image.bb]: Initially I had had this only in the minimal and and had hoped the the full image wouldn't need it but it proved to be wrong: it seems like this thing is bound to the recipe name. Using ${PN} here also shouldn't work but this is a guess that I'll try out tomorrow to be sure.

File: meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+IMAGE_INSTALL_append = " libstdc++"
+IMAGE_INSTALL_append = " u-boot-fw-utils"
+IMAGE_INSTALL_append = " file"
+IMAGE_INSTALL_append = " pciutils"
+IMAGE_INSTALL_append = " usbutils"
+IMAGE_INSTALL_append = " lshw"
+IMAGE_INSTALL_append = " ldd"
+IMAGE_INSTALL_append = " i2c-tools"
+IMAGE_INSTALL_append = " watchdog-sample"
+IMAGE_INSTALL_append = " systemd-analyze"
```
[the remark this answers is not in the document]
Comment (staroselskii) [reply meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb]: No need to add it the minimal image. Just left it here to discuss. I'll move it.

File: meta-intel-edison-distro/recipes-core/images/edison-image.bb [hunk 3]
```diff
@@ -84,16 +40,8 @@ IMAGE_INSTALL_append = " python"
 IMAGE_INSTALL_append = " python-argparse"
```
[the remark this answers is not in the document]
Comment (staroselskii) [reply meta-intel-edison-distro/recipes-core/images/edison-image.bb]: Thanks!

File: meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+IMAGE_INSTALL_append = " less"
+
+# Those are necessary to manually create partitions and file systems on the eMMC
+IMAGE_INSTALL_append = " parted"
+IMAGE_INSTALL_append = " e2fsprogs-e2fsck e2fsprogs-mke2fs e2fsprogs-tune2fs e2fsprogs-badblocks libcomerr libss libe2p libext2fs dosfstools"
+
+# Time related
+IMAGE_INSTALL_append = " tzdata"
+
+IMAGE_INSTALL_append = " libgpiod"
```
[the remark this answers is not in the document]
Comment (staroselskii) [reply meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb]: libgpiod is so useful that I'm inclined to leave it here if you're not against it.

File: meta-intel-edison-distro/recipes-core/images/edison-image.bb (hunk 1 above)
[the remark this answers is not in the document]
Comment (htot) [reply meta-intel-edison-distro/recipes-core/images/edison-image.bb]: Ah, I see. Maybe @alext-mkrs can advise here?

File: meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb (hunk 4 above)
[the remark this answers is not in the document]
Comment (htot) [reply meta-intel-edison-distro/recipes-core/images/edison-image-minimal.bb]: It is used in the initramfs, so I think should stay in minimal

**Comment by htot** [comment]
No rush.

File: meta-intel-edison-distro/recipes-core/images/edison-image.bb (hunk 1 above)
[the remark this answers is not in the document]
Comment (alext-mkrs) [reply meta-intel-edison-distro/recipes-core/images/edison-image.bb]: This is dependent on the recipe name by design of the overrides feature. I haven't tried using namely ${PN} (or so I think - it was a while ago), so feel free to give ita try. If that doesn't work, we'll need it in both.

**Comment by staroselskii** [comment]
@htot @alext-mkrs 

Thank you guys for the review. I just force pushed. It'd be nice if someone could build it before merging.

**Comment by htot** [comment]
Thanks. I'll test the full image prior to merging.

**Comment by htot** [comment]
Merged

**Pull request closed without merging**